archive/multisource.go

```diff
@@ -92,59 +91,120 @@ func NewMultiSource(ark *Archive, altPaths []string) driver.MultiSource {
 	if len(altPaths) == 1 && altPaths[0] == "_" {
 		altPaths = nil
 	}
-	return &multiSource{
-		ark:      ark,
-		altPaths: altPaths,
+	if len(altPaths) != 0 {
+		return &chunkMultiSource{
+			ark:      ark,
+			altPaths: altPaths,
+		}
+	}
+	return &spanMultiSource{
+		ark: ark,
 	}
 }
 
-func (m *multiSource) OrderInfo() (field.Static, bool) {
-	if len(m.altPaths) == 0 {
-		return field.New("ts"), m.ark.DataOrder == zbuf.OrderDesc
-	}
-	return nil, false
+type spanMultiSource struct {
+	ark *Archive
 }
 
-func (m *multiSource) spanWalk(ctx context.Context, zctx *resolver.Context, sf driver.SourceFilter, srcChan chan<- driver.SourceOpener) error {
-	return SpanWalk(ctx, m.ark, sf.Span, func(si SpanInfo) error {
-		so := func() (driver.ScannerCloser, error) {
-			return newSpanScanner(ctx, m.ark, zctx, sf.Filter, sf.FilterExpr, si)
-		}
+func (m *spanMultiSource) OrderInfo() (field.Static, bool) {
+	return field.New("ts"), m.ark.DataOrder == zbuf.OrderDesc
+}
+
+func (m *spanMultiSource) SendSources(ctx context.Context, span nano.Span, srcChan chan driver.Source) error {
+	return SpanWalk(ctx, m.ark, span, func(si SpanInfo) error {
 		select {
-		case srcChan <- so:
+		case srcChan <- &spanSource{ark: m.ark, spanInfo: si}:
 			return nil
 		case <-ctx.Done():
 			return ctx.Err()
 		}
 	})
 }
 
-func (m *multiSource) chunkWalk(ctx context.Context, zctx *resolver.Context, sf driver.SourceFilter, srcChan chan<- driver.SourceOpener) error {
-	return Walk(ctx, m.ark, func(chunk Chunk) error {
-		so := func() (driver.ScannerCloser, error) {
-			paths := make([]string, len(m.altPaths))
-			for i, input := range m.altPaths {
-				paths[i] = chunk.Localize(m.ark, input).String()
-			}
-			rc := detector.MultiFileReader(zctx, paths, zio.ReaderOpts{Format: "zng"})
-			sn, err := scanner.NewScanner(ctx, rc, sf.Filter, sf.FilterExpr, sf.Span)
-			if err != nil {
-				return nil, err
-			}
-			return &scannerCloser{Scanner: sn, Closer: rc}, nil
+func (m *spanMultiSource) SourceFromRequest(ctx context.Context, req *api.WorkerRequest) (driver.Source, error) {
+	si := SpanInfo{Span: req.Span}
+	for _, p := range req.ChunkPaths {
+		tsd, _, id, ok := parseChunkRelativePath(p)
+		if !ok {
+			return nil, zqe.E(zqe.Invalid, "invalid chunk path: %v", p)
+		}
+		md, err := readChunkMetadata(ctx, chunkMetadataPath(m.ark, tsd, id))
+		if err != nil {
+			return nil, err
 		}
+		si.Chunks = append(si.Chunks, md.Chunk(id))
+	}
+	return &spanSource{
+		ark:      m.ark,
+		spanInfo: si,
+	}, nil
+}
+
+type spanSource struct {
+	ark      *Archive
+	spanInfo SpanInfo
+}
+
+func (s *spanSource) Open(ctx context.Context, zctx *resolver.Context, sf driver.SourceFilter) (driver.ScannerCloser, error) {
+	return newSpanScanner(ctx, s.ark, zctx, sf.Filter, sf.FilterExpr, s.spanInfo)
+}
+
+func (s *spanSource) ToRequest(req *api.WorkerRequest) error {
```

Comment:
(This is more of a comment for further discussion, I don't think this should be addressed in this pr): I do not think archive (or driver for that matter) should import packages from zqd. packages outside of zqd should be lower level than zqd. At some point we should consider moving all logic having to do with running queries through distributed zqd instances into the zqd package. The apis in archive and driver should have hooks to make this possible.

Reply:
agreed; this was part of a commit I wrote to unblock the distributed demo work, and not intended to be permanent. @nwt , @marktwallace and I had a few discussions about how to structure this, but haven't yet decided on the best way, and thought this was ok for the moment.